File: security_cam/lib/detection/motion_detector.dart
import 'dart:typed_data';

import '../core/detector.dart';
import '../core/models.dart';
import 'regions/region_filter.dart';

class MotionDetector extends FrameDetector {
  @override
  final DetectorConfig config;
  static const _pixelDiffTolerance = 30;

  GrayscaleBitmap? _previous;
  int _persistenceCount = 0;
  Uint8List? _mask;
  int _maskCount = 0;
  int _maskWidth = 0;
  int _maskHeight = 0;
  List<DetectionRegion>? _maskRegions;

  MotionDetector(this.config);

  @override
  String get id => config.type;

  @override
  String get triggerType => TriggerType.motion;

  @override
  Future<void> init() async {}

  @override
  void reset() {
    _previous = null;
    _persistenceCount = 0;
    _mask = null;
  }

  @override
  Future<void> dispose() async {}

  @override
  DetectionResult analyzeFrame(AnalysisFrame frame) {
    if (_mask == null ||
        !identical(_maskRegions, regions) ||
        _maskWidth != frame.bitmap.width ||
        _maskHeight != frame.bitmap.height) {
      _rebuildMask(frame.bitmap.width, frame.bitmap.height);
    }
    final bitmap = frame.bitmap;
    final prev = _previous;
    _previous = bitmap;
    if (prev == null) {
      return _result(frame.timestamp, 0.0, false);
    }
    final ratio = _diffRatio(prev, bitmap);
    final triggered = _updatePersistence(ratio, frame.timestamp);
    return _result(frame.timestamp, ratio, triggered);
  }

  void _rebuildMask(int width, int height) {
    final (mask, count) = pixelMask(regions, width, height);
    _mask = mask;
    _maskCount = count;
    _maskRegions = regions;
    _maskWidth = width;
    _maskHeight = height;
  }

  bool _updatePersistence(double ratio, DateTime timestamp) {
    final above = ratio >= config.threshold;
    _persistenceCount = above ? _persistenceCount + 1 : 0;
    if (_persistenceCount >= config.persistenceFrames) {
      _persistenceCount = 0;
      return true;
    }
    return false;
  }

  double _diffRatio(GrayscaleBitmap a, GrayscaleBitmap b) {
    final mask = _mask!;
    final count = _maskCount;
    var changed = 0;
    for (var y = 0; y < a.height; y++) {
      final rowA = y * a.width;
      final rowB = y * a.width;
      for (var x = 0; x < a.width; x++) {
        final idx = rowA + x;
        if (mask[idx] == 0) continue;
        final diff = (a.gray[idx] - b.gray[idx]).abs();
        if (diff > _pixelDiffTolerance) changed++;
      }
    }
    return count == 0 ? 0.0 : changed / count;
  }

  DetectionResult _result(DateTime ts, double score, bool triggered) {
    return DetectionResult(
      timestamp: ts,
      triggerType: triggerType,
      score: score,
      triggered: triggered,
    );
  }
}

Uint8List buildFrame(int width, int height, int fill) {
  return Uint8List(width * height)..fillRange(0, width * height, fill);
}

Uint8List buildFrameWithRect(
    int width, int height, int fill, int rectX, int rectY, int rectW, int rectH, int rectFill) {
  final buf = Uint8List(width * height)..fillRange(0, width * height, fill);
  for (var y = rectY; y < rectY + rectH && y < height; y++) {
    for (var x = rectX; x < rectX + rectW && x < width; x++) {
      buf[y * width + x] = rectFill;
    }
  }
  return buf;
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/detection/MotionDetector.kt
package io.securitycam.level1.detection

import io.securitycam.level1.core.TriggerType
import java.time.Instant

private const val PIXEL_DIFF_TOLERANCE = 30

/**
 * Grayscale pixel-diff motion detector (port of
 * `lib/detection/motion_detector.dart`). First frame primes; consecutive
 * above-threshold ratio frames for [DetectorConfig.persistenceFrames] trigger.
 */
class MotionDetector(
    override val config: io.securitycam.level1.detection.DetectorConfig,
) : FrameDetector() {

    private var previous: GrayscaleBitmap? = null
    private var persistenceCount = 0
    private var mask: ByteArray? = null
    private var maskCount = 0
    private var maskWidth = 0
    private var maskHeight = 0
    private var maskRegions: List<DetectionRegion>? = null

    override val id: String get() = config.type
    override val triggerType: String get() = TriggerType.motion

    override suspend fun init() {}

    override fun reset() {
        previous = null
        persistenceCount = 0
        mask = null
    }

    override suspend fun dispose() {}

    override fun analyzeFrame(frame: AnalysisFrame): DetectionResult {
        if (mask == null ||
            maskRegions !== regions ||
            maskWidth != frame.bitmap.width ||
            maskHeight != frame.bitmap.height
        ) {
            rebuildMask(frame.bitmap.width, frame.bitmap.height)
        }
        val bitmap = frame.bitmap
        val prev = previous
        previous = bitmap
        if (prev == null) {
            return result(frame.timestamp, 0.0, false)
        }
        val ratio = diffRatio(prev, bitmap)
        val triggered = updatePersistence(ratio, frame.timestamp)
        return result(frame.timestamp, ratio, triggered)
    }

    private fun rebuildMask(width: Int, height: Int) {
        val (newMask, count) = RegionFilter.pixelMask(regions, width, height)
        mask = newMask
        maskCount = count
        maskRegions = regions
        maskWidth = width
        maskHeight = height
    }

    private fun updatePersistence(ratio: Double, timestamp: Instant): Boolean {
        val above = ratio >= config.threshold
        persistenceCount = if (above) persistenceCount + 1 else 0
        if (persistenceCount >= config.persistenceFrames) {
            persistenceCount = 0
            return true
        }
        return false
    }

    private fun diffRatio(a: GrayscaleBitmap, b: GrayscaleBitmap): Double {
        val mask = mask ?: return 0.0
        val count = maskCount
        var changed = 0
        for (y in 0 until a.height) {
            val rowA = y * a.width
            for (x in 0 until a.width) {
                val idx = rowA + x
                if (mask[idx].toInt() == 0) continue
                val diff = (a.gray[idx].toInt() and 0xFF) - (b.gray[idx].toInt() and 0xFF)
                if (kotlin.math.abs(diff) > PIXEL_DIFF_TOLERANCE) changed++
            }
        }
        return if (count == 0) 0.0 else changed.toDouble() / count
    }

    private fun result(ts: Instant, score: Double, triggered: Boolean): DetectionResult =
        DetectionResult(
            timestamp = ts,
            triggerType = triggerType,
            score = score,
            triggered = triggered,
        )
}

/** Test helper: uniform gray frame. */
fun buildFrame(width: Int, height: Int, fill: Int): ByteArray {
    val buf = ByteArray(width * height)
    buf.fill(fill.toByte())
    return buf
}

/** Test helper: frame with a filled rectangle. */
fun buildFrameWithRect(
    width: Int, height: Int, fill: Int,
    rectX: Int, rectY: Int, rectW: Int, rectH: Int, rectFill: Int,
): ByteArray {
    val buf = buildFrame(width, height, fill)
    for (y in rectY until rectY + rectH) {
        if (y >= height) break
        for (x in rectX until rectX + rectW) {
            if (x >= width) break
            buf[y * width + x] = rectFill.toByte()
        }
    }
    return buf
}